Mathlib/Analysis/InnerProductSpace/Quotient.lean

```diff
@@ -0,0 +1,201 @@
+/-
+Copyright (c) 2024 Yoh Tanimoto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yoh Tanimoto
+-/
+import Mathlib.Analysis.InnerProductSpace.Dual
+import Mathlib.Analysis.Normed.Group.Quotient
+
+/-!
+# Canonial inner product space from Preinner product space
+
+This file defines the inner product space from a preinner product space (the inner product
+can be degenerate) by quotienting the null space.
+
+## Main results
+
+-/
+
+noncomputable section
+
+open RCLike Submodule Quotient LinearMap InnerProductSpace InnerProductSpace.Core
+
+variable (𝕜 E : Type*) [k: RCLike 𝕜]
+
+section Nullspace
+
+variable [SeminormedAddCommGroup E] [InnerProductSpace 𝕜 E]
+
+/-- The null space with respect to the canonical inner product. It is defined by `‖x‖ = 0` and
+it is proven using the Cauchy-Schwarz inequality that ` ⟪x, y⟫_𝕜 = 0` for all `y : E`. -/
```

Comment:
This isn't true, cauchy-schwarz is not used in this definition at all!

Note you can define this more generally as
```lean
variable {𝕜 E} [SeminormedRing 𝕜] [SeminormedAddCommGroup E] [Module 𝕜 E] [BoundedSMul 𝕜 E]

/-- The null space with respect to the norm. -/
def nullSpace : Submodule 𝕜 E where
  carrier := {x : E | ‖x‖ = 0}
  add_mem' {x y} (hx : ‖x‖ = 0) (hy : ‖y‖ = 0) := by
    apply le_antisymm _ (norm_nonneg _)
    refine (norm_add_le x y).trans_eq ?_
    rw [hx, hy, add_zero]
  zero_mem' := norm_zero
  smul_mem' c x (hx : ‖x‖ = 0) := by
    apply le_antisymm _ (norm_nonneg _)
    refine (norm_smul_le _ _).trans_eq ?_
    rw [hx, mul_zero]
```

Reply:
I see, I will move the docstring to the beginning.

I get "failed to synthesize `BoundedSMul 𝕜 E`", although it is there... why?
```lean4
import Mathlib.Analysis.InnerProductSpace.Dual
import Mathlib.Analysis.Normed.Group.Quotient

variable (𝕜 E : Type*) [k: RCLike 𝕜]

variable {𝕜 E} [SeminormedRing 𝕜] [SeminormedAddCommGroup E] [Module 𝕜 E] [BoundedSMul 𝕜 E]

/-- The null space with respect to the norm. -/
def nullSpace : Submodule 𝕜 E where
  carrier := {x : E | ‖x‖ = 0}
  add_mem' {x y} (hx : ‖x‖ = 0) (hy : ‖y‖ = 0) := by
    apply le_antisymm _ (norm_nonneg _)
    refine (norm_add_le x y).trans_eq ?_
    rw [hx, hy, add_zero]
  zero_mem' := norm_zero
  smul_mem' c x (hx : ‖x‖ = 0) := by
    apply le_antisymm _ (norm_nonneg _)
    refine (norm_smul_le _ _).trans_eq ?_
    rw [hx, mul_zero]
```

Reply:
Ah ok, thank you!

I think I will have to wait until I will have decided to use `SeparationQuotient` or the null space by hand.